conjure-java-jaxrs-client/src/main/java/com/palantir/conjure/java/client/jaxrs/DialogueFeignClient.java

```diff
@@ -0,0 +1,346 @@
+import com.palantir.dialogue.ConjureRuntime;
+import com.palantir.dialogue.Deserializer;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.HttpMethod;
+import com.palantir.dialogue.RequestBody;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.UrlBuilder;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import feign.Request;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.Reader;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * {@link DialogueFeignClient} is an adapter from {@link feign.Client} to {@link Channel Dialogue Channel}
+ * taking advantage of the superior observability and stability provided by Dialogue.
+ */
+final class DialogueFeignClient implements feign.Client {
+
+    private static final String PATH_TEMPLATE = "hr-path-template";
+    private static final Splitter pathSplitter = Splitter.on('/').omitEmptyStrings();
+    private static final Splitter.MapSplitter querySplitter =
+            Splitter.on('&').omitEmptyStrings().withKeyValueSeparator('=');
+
+    private final Clients clients;
+    private final Channel channel;
+    private final String baseUrl;
+    private final String serviceName;
+    private final String version;
+
+    DialogueFeignClient(Class<?> jaxrsInterface, Channel channel, ConjureRuntime runtime, String baseUrl) {
+        this.channel = Preconditions.checkNotNull(channel, "Channel is required");
+        this.serviceName = jaxrsInterface.getSimpleName();
+        this.version = Optional.ofNullable(jaxrsInterface.getPackage().getImplementationVersion())
+                .orElse("0.0.0");
+        this.baseUrl = baseUrl;
+        this.clients = runtime.clients();
+    }
+
+    @Override
+    public feign.Response execute(Request request, Request.Options _options) throws IOException {
+        com.palantir.dialogue.Request.Builder builder =
+                com.palantir.dialogue.Request.builder().body(requestBody(request));
+        request.headers().forEach((headerName, values) -> {
+            if (includeRequestHeader(headerName)) {
+                builder.putAllHeaderParams(headerName, values);
+            }
+        });
+
+        try {
+            Response response = clients.block(clients.call(
+                    channel, new FeignRequestEndpoint(request), builder.build(), IdentityDeserializer.INSTANCE));
+            return feign.Response.create(
+                    response.code(),
+                    null,
+                    Multimaps.asMap((Multimap<String, String>) response.headers()),
```

Comment:
why the casting here? couldn't we just do `response.headers().asMap()` ?

Reply:
ListMultimap.asMap returns `Map<K, List<V>>`, by casting to Multimap we get the requested `Map<K, Collection<V>>`